File: frontend/pages/Feedback_Page/Feedback_Page.py
"""
Feedback_Page.py
--------------
Description:
    Handles user feedback collection for the Geometric Learning System.
    Manages feedback form display and submission processing.

Author: Karin Hershko and Afik Dadon
Date: February 2025
"""

from flask import Blueprint, render_template, request, redirect, url_for, session, jsonify
from db_utils import get_db_connection
from UserLogger import UserLogger
from typing import Dict, Union
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

feedback_page = Blueprint('feedback_page', __name__,
                          template_folder='templates',
                          static_folder='static')


@feedback_page.route('/')
def feedback() -> Union[str, redirect]:
    """Display the feedback form page.
    Requires user authentication."""
    user = session.get('user', None)
    if not user:
        return redirect(url_for('login_page.login'))
    return render_template('Feedback_Page.html')


@feedback_page.route('/submit', methods=['POST'])
def submit_feedback() -> jsonify:
    """Process feedback form submission.
    Validates input and stores in database."""
    try:
        # Validate user session
        if not _validate_user_session():
            return jsonify({
                'success': False,
                'error': 'User is not logged in'
            }), 401

        # Get and validate form data
        data = request.get_json()
        if not data:
            logger.error("No feedback data received")
            return jsonify({
                'success': False,
                'error': 'No feedback data received'
            }), 400

        # Save feedback to database
        _save_feedback_to_db(data)

        # Log successful submission
        UserLogger.log_feedback_submission()

        return jsonify({'success': True})

    except Exception as e:
        logger.error(f"Error in submit_feedback: {str(e)}")
        return jsonify({
            'success': False,
            'error': 'אירעה שגיאה בשמירת המשוב. אנא נסה שוב או צור קשר עם התמיכה.'
        }), 500


# === Helper Functions ===

def _validate_user_session() -> bool:
    """Validate user session and extract user ID."""
    user = session.get('user')
    if not user:
        logger.warning("No user in session during feedback submission")
        return False

    user_id = user.get('user_id')
    if not user_id:
        logger.warning("No user ID in session")
        return False

    return True


def _save_feedback_to_db(data: Dict) -> None:
    """Save feedback data to database."""
    user_id = session['user']['user_id']

    with get_db_connection() as conn:
        cursor = conn.cursor()

        query = """
            INSERT INTO UserFeedback (
                user_id, 
                usability_easy_to_use, usability_clear_questions,
                usability_clear_interface, usability_easy_navigation,
                educational_concepts, educational_theorems,
                educational_guidance, educational_learning,
                format_dont_know_helpful, format_sufficient_options,
                format_would_use_again,
                intelligence_understood_responses, intelligence_relevant_questions,
                missing_questions, unclear_questions,
                suggested_improvements, expected_questions
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """

        values = (
            user_id,
            data.get('usability_easy_to_use'),
            data.get('usability_clear_questions'),
            data.get('usability_clear_interface'),
            data.get('usability_easy_navigation'),
            data.get('educational_concepts'),
            data.get('educational_theorems'),
            data.get('educational_guidance'),
            data.get('educational_learning'),
            data.get('format_dont_know_helpful'),
            data.get('format_sufficient_options'),
            data.get('format_would_use_again'),
            data.get('intelligence_understood_responses'),
            data.get('intelligence_relevant_questions'),
            data.get('missing_questions', ''),
            data.get('unclear_questions', ''),
            data.get('suggested_improvements', ''),
            data.get('expected_questions', '')
        )

        cursor.execute(query, values)

File: frontend/pages/Feedback_Page/static/js/Feedback_Page.js
/**
 * Feedback_Page.js
 * --------------
 * Client-side functionality for the feedback form including:
 * - Form submission handling
 * - Input validation
 * - Error handling
 * - UI state management
 *
 * Karin Hershko and Afik Dadon
 * Date: February 2025
 */

document.addEventListener('DOMContentLoaded', function() {
    // === DOM Elements ===
    const form = document.getElementById('feedbackForm');
    const successMessage = document.querySelector('.success-message');
    const errorMessage = document.querySelector('.error-message');
    const submitButton = form.querySelector('.submit-btn');

    // Verify required elements
    if (!form) {
        console.error('Feedback form not found');
        return;
    }

    // === Event Listeners ===
    form.addEventListener('submit', handleFormSubmit);
    initializeValidation();

    /**Handle form submission
     * @param {Event} e - Submit event */
    async function handleFormSubmit(e) {
        e.preventDefault();
        resetMessages();
        setLoadingState(true);

        try {
            const formData = collectFormData();
            const response = await submitFeedback(formData);
            await handleSubmissionResponse(response);
        } catch (error) {
            handleSubmissionError(error);
        } finally {
            setLoadingState(false);
        }
    }

    // === Form Data Handling ===

    /**Collect all form data into an object
     * @returns {Object} Collected form data*/
    function collectFormData() {
        return {
            // Usability questions
            usability_easy_to_use: getRadioValue('usability_easy_to_use'),
            usability_clear_questions: getRadioValue('usability_clear_questions'),
            usability_clear_interface: getRadioValue('usability_clear_interface'),
            usability_easy_navigation: getRadioValue('usability_easy_navigation'),

            // Educational value
            educational_concepts: getRadioValue('educational_concepts'),
            educational_theorems: getRadioValue('educational_theorems'),
            educational_guidance: getRadioValue('educational_guidance'),
            educational_learning: getRadioValue('educational_learning'),

            // Format questions
            format_dont_know_helpful: getRadioValue('format_dont_know_helpful'),
            format_sufficient_options: getRadioValue('format_sufficient_options'),
            format_would_use_again: getRadioValue('format_would_use_again'),

            // System intelligence
            intelligence_understood_responses: getRadioValue('intelligence_understood_responses'),
            intelligence_relevant_questions: getRadioValue('intelligence_relevant_questions'),

            // Open questions
            missing_questions: getTextareaValue('missing_questions'),
            unclear_questions: getTextareaValue('unclear_questions'),
            suggested_improvements: getTextareaValue('suggested_improvements'),
            expected_questions: getTextareaValue('expected_questions')
        };
    }

    /**Submit feedback data to server
     * @param {Object} formData - Collected form data
     * @returns {Promise} Fetch response*/
    async function submitFeedback(formData) {
        return await fetch('/feedback/submit', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(formData),
            credentials: 'include'
        });
    }

    // === UI State Management ===

    /** Set loading state for form
     * @param {boolean} isLoading - Whether form is in loading state*/
    function setLoadingState(isLoading) {
        if (submitButton) {
            submitButton.textContent = isLoading ? 'שולח...' : 'שלח משוב';
            submitButton.disabled = isLoading;
            submitButton.classList.toggle('loading', isLoading);
        }
    }

    /**Reset all message displays*/
    function resetMessages() {
        if (successMessage) successMessage.style.display = 'none';
        if (errorMessage) errorMessage.style.display = 'none';
    }

    /**Handle successful form submission*/
    async function handleSubmissionResponse(response) {
        const result = await response.json();

        if (result.success) {
            form.reset();
            showThankYouModal();
            setTimeout(redirectToHome, 3000);
        } else {
            throw new Error(result.error || 'Failed to submit feedback');
        }
    }

    /**Display thank you modal*/
    function showThankYouModal() {
        const modal = document.getElementById('thankYouModal');
        if (modal) {
            modal.style.display = 'flex';
        }
    }

    /**Redirect to home page*/
    function redirectToHome() {
        window.location.href = '/';
    }

    /**Handle submission error
     * @param {Error} error - Error object*/
    function handleSubmissionError(error) {
        console.error('Submission error:', {
            message: error.message,
            stack: error.stack
        });

        displayErrorMessage(error.message);
    }

    /**Display error message to user
     * @param {string} message - Error message to display*/
    function displayErrorMessage(message) {
        if (!errorMessage) {
            const newErrorMessage = createErrorMessage(message);
            form.parentNode.insertBefore(newErrorMessage, form.nextSibling);
        } else {
            errorMessage.style.display = 'block';
            errorMessage.textContent = 'אירעה שגיאה בשליחת המשוב: ' + message;
        }
    }

    /**Create new error message element
     * @param {string} message - Error message
     * @returns {HTMLElement} Error message element*/
    function createErrorMessage(message) {
        const element = document.createElement('div');
        element.className = 'error-message';
        element.textContent = 'אירעה שגיאה בשליחת המשוב: ' + message;
        return element;
    }

    // === Form Validation ===

    /**Initialize form validation*/
    function initializeValidation() {
        document.querySelectorAll('input[required]').forEach(input => {
            input.addEventListener('invalid', handleInvalidInput);
        });
    }

    /**Handle invalid input
     * @param {Event} e - Invalid event*/
    function handleInvalidInput(e) {
        e.preventDefault();
        this.closest('.feedback-section').scrollIntoView({
            behavior: 'smooth',
            block: 'start'
        });
    }

    // === Helper Functions ===

    /**Get value of selected radio button
     * @param {string} name - Radio button name
     * @returns {string|null} Selected value or null*/
    function getRadioValue(name) {
        const radio = document.querySelector(`input[name="${name}"]:checked`);
        return radio ? radio.value : null;
    }

    /**Get textarea value
     * @param {string} id - Textarea ID
     * @returns {string} Textarea value*/
    function getTextareaValue(id) {
        const textarea = document.getElementById(id);
        return textarea ? textarea.value.trim() : '';
    }
});